src/reflection.ts

```diff
@@ -802,18 +811,24 @@ export class Parser {
           continue;
         }
 
-        const deserializer = unionDeserializers.get(discriminator);
-        if (!deserializer) {
+        const specificTypeInfo = specificTypes.get(discriminator);
+        if (!specificTypeInfo) {
           throw new Error(
             `Malformed message: unknown union type '${discriminator}' in field ${discriminatorField.name()}`,
           );
         }
 
-        const subTable = tables[idx];
-        if (!subTable) {
-          throw new Error(`Malformed message missing table at ${field.name()} positon ${idx}`);
-        }
-        result.push(deserializer(subTable));
+        const subTable = new Table(
+          table.bb,
+          specificTypeInfo.typeIndex,
+          table.bb.__indirect(nextTableOffset),
+          false /* elementIsStruct */,
```

Comment:
I too suspect it would not be crazy - but I'm gonna vote we punt on this to a follow-up PR. I can take it on that but there's less urgency than this bugfix.

Reply:
Absolutely punt it to later (if ever).